Enhance `useAdsCurrency` tests and add a case for testing the fallback values of currency config

eason9487 opened the pull request:
### Changes proposed in this Pull Request:

I was eager to fix the bug first yesterday, so I did not add the corresponding test with #1252. This PR makes up for the test.

- Add a test case for `useAdsCurrency` hook to check the fallback values of currency config.

:warning: **UPDATE: this PR replaced the mocking from `apiFetch` to `useGoogleAdsAccount`.** Please refer to https://github.com/woocommerce/google-listings-and-ads/pull/1255#discussion_r809977404 for more context.

### Detailed test instructions:

1. `npm run test:js -- useAdsCurrency`
2. All tests should pass.
3. If assign [default values](https://github.com/woocommerce/google-listings-and-ads/blob/889d3396c79a6a102f78df0f523b5d8b4af0fe4e/js/src/hooks/useAdsCurrency.js#L49-L50) with destructuring assignment such as the following, the test should find out the problem.
    ```js
    const { currency: code = '', symbol = '' } = googleAdsAccount || {};
    ```
    (Got from the fix commit: https://github.com/woocommerce/google-listings-and-ads/pull/1252/commits/1388b515e487cafe52aecb7472b4caebc73bb8e4)

### Changelog entry

eason9487 committed:
Add test case for useAdsCurrency to check the fallback config with unconnected Google Ads account

jorgemd24 commented on js/src/hooks/useAdsCurrency.test.js:
```diff
@@ -92,10 +94,15 @@ describe( 'useAdsCurrency', () => {
+				// Mock /wp-json/wc/gla/ads/connection as a connected Ads account
+				.mockReturnValueOnce( adsAccountData )
+				// Mock /wp-json/wc/gla/ads/connection as a unconnected Ads account
+				.mockReturnValueOnce( {
+					id: 0,
+					currency: null,
+					symbol: null,
+					status: 'disconnected',
+				} );
```
❓ 

No introduced in this PR but It makes it a bit difficult to follow how the mocks are returned as it depends on the calls on `apiFetch` instead of the actual value returned from the hook. 

For example, that section of code can be omitted and the test will still pass as the properties currency and symbol will not exist therefore the value will be an empty string. I think if we really want to test the case where currency and symbol are `null` we should mock the actual value returned from useGoogleAdsAccount. 

Something that I noticed was that when I want to run a specific test with `test.only` the `googleAdsAccount` always is resolved as undefined.

 `const { googleAdsAccount, hasFinishedResolution } = useGoogleAdsAccount();`

eason9487 replied:
I asked a relevant question before - https://github.com/woocommerce/google-listings-and-ads/pull/920#discussion_r689447682. I guess in addition to the purpose mentioned by @tomalec, it was also to verify the `hasFinishedResolution` value with its changes.

I did feel the problems you mentioned when I added its tests recently, and the test was difficult to write. Moreover, when its tests fail, the messages displayed are also confusing. It only shows timed out and devs have to do more local changes to find out which assertions are not passed.
![image](https://user-images.githubusercontent.com/17420811/154894589-8e72253c-f6ca-497a-bcf3-c22cb5f6a2a3.png)

After looking back at the codebase, I noticed that the `hasFinishedResolution` property in the returned object of `useAdsCurrency` is not used anywhere. And it also handles the fallback values for currency config before/after the Google Ads account status is fetched. I'd like to propose to remove `hasFinishedResolution` and test it by mocking `useGoogleAdsAccount` instead. In this way, the pros would be

1. Tests are atomic.
2. Easier to write and read tests.
3. The error messages are clearer.
    ![image](https://user-images.githubusercontent.com/17420811/154895726-f80c48b5-ee0c-4dc8-a043-7f2b6d90fc48.png)

💡  The commit https://github.com/woocommerce/google-listings-and-ads/commit/fab8e78f352196b528b4bf12e70e936a18c54928 is the proposed change. Please have a look.

---

Or, to keep the original purpose and make the tests to be atomic and the error messages clearer, we could change tests in this way

```js
import { renderHook, act } from '@testing-library/react-hooks';
import { apiFetch } from '@wordpress/data-controls';
import { dispatch, select, subscribe } from '@wordpress/data';

import { STORE_KEY } from '.~/data';

describe( 'useAdsCurrency', () => {
  describe( 'eventually', () => {
    let prepareAdsAccount;

    beforeEach( () => {
      apiFetch
        // Mock /wp-json/wc/gla/jetpack/connected
        .mockReturnValueOnce( {
          active: 'yes',
          owner: 'yes',
          displayName: 'username',
          email: 'tomalec.gla1.test@gmail.com',
        } )
        // Mock /wp-json/wc/gla/google/connected
        .mockReturnValueOnce( {
          active: 'yes',
          email: 'gla1.test@example.com',
        } );

      // Each tests prepare the mocking ads account data independently so that they can be atomic.
      prepareAdsAccount = ( nextAdsAccountData ) => {
        // Mock /wp-json/wc/gla/ads/connection
        apiFetch.mockReturnValueOnce( nextAdsAccountData );

        dispatch( STORE_KEY ).invalidateResolution(
          'getGoogleAdsAccount',
          []
        );

        return new Promise( ( resolve ) => {
          subscribe( () => {
            const { hasFinishedResolution } = select( STORE_KEY );
            if ( hasFinishedResolution( 'getGoogleAdsAccount' ) ) {
              resolve();
            }
          } );
        } );
      };
    } );

    // ...omitted

    test( 'when Ads account is not connected, it should return the code/symbol of currency config with fallback value', async () => {
      const { result } = renderHook( () => useAdsCurrency() );

      // Use `act` to wait for the data fetching of wp-data instead of `waitFor`
      await act( () =>
        prepareAdsAccount( {
          id: 0,
          currency: null,
          symbol: null,
          status: 'disconnected',
        } )
      );

      const { adsCurrencyConfig, formatAmount } = result.current;
      const value = 1234.5678;

      expect( adsCurrencyConfig.code ).toBe( '' );
      expect( adsCurrencyConfig.symbol ).toBe( '' );
      expect( formatAmount( value ) ).toBe( '1|234;568\xA0' );
      expect( formatAmount( value, true ) ).toBe( '1|234;568\xA0' );
    } );
  } );
} );
```

---

I lean towards the mock `useGoogleAdsAccount` solution because it should be equally effective in this codebase as well and be easier to understand and write. What do you think?

jorgemd24 replied:
Thanks @eason9487 to point out that comment, It helps to understand what is the reason for mocking the apiFetch instead of mocking the hooks. I agree with you that makes the implementation a lot harder and more difficult to trace the mock values. I prefer the first option, mocking useGoogleAdsAccount and I see in [fab8e78](https://github.com/woocommerce/google-listings-and-ads/commit/fab8e78f352196b528b4bf12e70e936a18c54928) it is a lot easier to read it and it makes the test more robust. 

The second option I think it gets too complex and more fragile, what would it happen if one of the hooks are adding a new call to apiFetch? The good point of this approach is that the test will reflect a more realistic scenario (from the behaviour perspective) as is mentioned in https://github.com/woocommerce/google-listings-and-ads/pull/920#discussion_r689447682  .

eason9487 replied:
> The second option I think it gets too complex and more fragile, what would it happen if one of the hooks are adding a new call to apiFetch?

It's okay to add more calls to `apiFetch` within a `test()` via `prepareAdsAccount`. But the complexity is it depends on the internal implementation and behaviors of wp-data, it's still hard to maintain.

If the new call is added into the shared part of `prepareAdsAccount`, it's still easy to cause unintended side effects or even break other tests.

> The good point of this approach is that the test will reflect a more realistic scenario (from the behaviour perspective) as is mentioned in [#920 (comment)](https://github.com/woocommerce/google-listings-and-ads/pull/920#discussion_r689447682) .

I think perhaps this could still be approached conceptually by adding testing for `useGoogleAccount` and `useGoogleAdsAccount` later. 🤔 

---

Thanks for the review comments! I will merge https://github.com/woocommerce/google-listings-and-ads/commit/fab8e78f352196b528b4bf12e70e936a18c54928 back to this PR.

jorgemd24 reviewed (commented):
In general, the tests are passing, however, I left a comment to see if we could improve it.

eason9487 committed:
Enhance the tests of `useAdsCurrency` hook.

- Mock `useGoogleAdsAccount`  instead of `apiFetch` in the useAdsCurrency.test.js
- Remove `hasFinishedResolution` from returned property of useAdsCurrency hook.

Proposal of https://github.com/woocommerce/google-listings-and-ads/pull/1255#discussion_r809977404

jorgemd24 commented on js/src/hooks/useAdsCurrency.js:
```diff
@@ -37,15 +36,15 @@ import useGoogleAdsAccount from './useGoogleAdsAccount';
-	const { googleAdsAccount, hasFinishedResolution } = useGoogleAdsAccount();
+	const { googleAdsAccount } = useGoogleAdsAccount();
 
 	// Apply store's foramtting config with the Ad's currency and symbol.
```
💅

typo - foramtting -> formatting

eason9487 replied:
Thanks! Fixed in https://github.com/woocommerce/google-listings-and-ads/pull/1255/commits/5d66532b073a4be2b890481ba06f32e5735f26a6.

jorgemd24 reviewed (approved):
LGTM 👍 

Thanks for the new implementation. I just left a comment about a minor typo in the useAdsCurrency hook.

eason9487 committed:
Fix a typo in useAdsCurrency.js

Address https://github.com/woocommerce/google-listings-and-ads/pull/1255#discussion_r811773543
